File: LeetCode/125.Valid_Palindrome/125.Valid_Palindrome.swift
class Solution {
    func isPalindrome(_ s: String) -> Bool {
        let chars = Array(s)
        var left = 0
        var right = chars.count - 1

        while left < right {
            while left < right && !chars[left].isLetter && !chars[left].isNumber {
                left += 1
            }
            while left < right && !chars[right].isLetter && !chars[right].isNumber {
                right -= 1
            }

            if chars[left].lowercased() != chars[right].lowercased() {
                return false
            }

            left += 1
            right -= 1
        }

        return true
    }
}

let solution = Solution()
let s = "A man, a plan, a canal: Panama"
print(solution.isPalindrome(s))

File: LeetCode/125.Valid_Palindrome/125.Valid_Palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        StringBuilder sb = new StringBuilder();
        String lowerS = s.toLowerCase();
        for (int i = 0; i < lowerS.length(); i++) {
            char ch = lowerS.charAt(i);
            if (Character.isLetterOrDigit(ch)) {
                sb.append(ch);
            }
        }
        String str = sb.toString();
        int left = 0, right = str.length() - 1;
        while (left <= right) {
            if (str.charAt(left) == str.charAt(right)) {
                left += 1;
                right -= 1;
            } else {
                return false;
            }
        }
        return true;
    }
}